### PR TITLE
Added 'ago' and 'till' pipe and the required format pipes

### PR DESCRIPTION
**Pull request by OdfFhem**
- 'ago' calcs the elapsed time since a dateTime value.
- 'till' calcs the time to reach a dateTime value.
- 'timeFormat' allows to format a time - given in seconds ('s') or milliseconds (',s').
- 'dd' means total days for the calculated time.
- 'hhhh' means total hours for the calculated time; 'hh' and 'h' means the hours for the remaining time of a day.
- 'mmmmmm' means total minutes for the calculated time; 'mm' and 'm' means the minutes for the remaining time of a hour.
- 'ssssssss' means total seconds for the calculated time; 'ss' and s' means the seconds for the remaining time of a minute.
- two replacement modes ('lower' and 'upper') allow to control the replacement process.

### CODE REVIEW

File: modules/ftui/ftui.helper.js [hunk 1]
```diff
@@ -227,6 +227,61 @@ export function durationFromSeconds(time) {
   return ret;
 }
 
+export function agoTillFormat(ms, format, mode='lower') {
+  var x = ms / 1000;
+
+  var SECONDS = Math.floor(x);
```
Comment (knowthelist) [inline modules/ftui/ftui.helper.js:233]: var is deprecated. Please use let or const instead.

File: examples/label.html [hunk 2]
```diff
@@ -80,6 +80,30 @@
...
+      <ftui-label [text]="ftuitest:dateTime4ago | toDate() | ago() | agoFormat('dd days')"></ftui-label>
+    </ftui-grid-tile>
+
+    <ftui-grid-tile row="3" col="7" height="2" width="2" shape="round">
+      <header>... till</header>
+      <ftui-label [text]="ftuitest:dateTime4till"></ftui-label>
+      <ftui-label [text]="ftuitest:dateTime4till | toDate() | till()"></ftui-label>
+      <ftui-label [text]="ftuitest:dateTime4till | toDate() | till() | tillFormat('[ dd.hh:mm:ss ]')"></ftui-label>
+      <ftui-label [text]="ftuitest:dateTime4till | toDate() | till() | tillFormat('[ still dd days + hh:mm:ss ]')"></ftui-label>
+      <ftui-label [text]="ftuitest:dateTime4till | toDate() | till() | tillFormat('%SSSSSSSSs','upper')"></ftui-label>
```
Comment (knowthelist) [inline examples/label.html:101]: These parameters are not very intuitive

Comment (OdfFhem) [reply examples/label.html]: The ambiguities have been removed.

File: modules/ftui/ftui.helper.js (hunk 1 above)
Comment (OdfFhem) [reply modules/ftui/ftui.helper.js]: 'var' replaced by 'const' or 'let'.
'Math.floor' replaced by '~~'.

File: modules/ftui/ftui.binding.js [hunk 3]
```diff
@@ -83,7 +102,7 @@ export class FtuiBinding {
       const filteredValue = this.filter(attributeValue, options.filter);
       const value = String(options.value).replace(/\$value/g, filteredValue);
       const [parameterId, deviceName, readingName] = ftuiHelper.parseReadingId(readingId);
-      console.log(options.cmd + '#' + deviceName+ '#' + readingName+ '#' + value)
+      console.log(options.cmd + '#' + deviceName + '#' + readingName + '#' + value)
```
Comment (knowthelist) [inline modules/ftui/ftui.binding.js:105]: please remove all console.log before commit changes, even if I accidentally left it in.

Comment (OdfFhem) [reply modules/ftui/ftui.binding.js]: Removed - it was the only one in this module.

File: modules/ftui/ftui.binding.js [hunk 4]
```diff
@@ -1,7 +1,26 @@
...
+const format = value => input => ftuiHelper.dateFormat(input, value);
+const round = value => input => ftuiHelper.round(input, value);
+const add = value => input => input + value;
+const multiply = value => input => input * value;
+const replace = (find, replace) => input => String(input).replace(find, replace);
+const map = value => input => ftuiHelper.getMatchingValue(parseHocon(value, true), input);
+const scale = (minIn, maxIn, minOut, maxOut) => input => ftuiHelper.scale(input, minIn, maxIn, minOut, maxOut);
+const ago = () => input => ftuiHelper.dateAgo(input);
+const till = () => input => ftuiHelper.dateTill(input);
+const timeFormat = (format,imode='ms',fmode='lower') => input => ftuiHelper.timeFormat(input, format, imode, fmode);
```
Comment (knowthelist) [inline modules/ftui/ftui.binding.js:19]: Please no abbreviations in variable names. What does imode mean? internalMode, inMode? fmode: formatMode? imode: inMode?

Comment (OdfFhem) [reply modules/ftui/ftui.binding.js:19]: imode means inputMode.
fmode means formatMode.

Replace ?

File: modules/ftui/ftui.helper.js [hunk 5]
```diff
@@ -227,6 +226,66 @@ export function durationFromSeconds(time) {
   return ret;
 }
 
+export function timeFormat(ms, format, imode='ms', fmode='lower') {
+  // imode: 'ms' or 's'
+  let x = ms
+  if (imode == 'ms') { x /= 1000; }
```
Comment (knowthelist) [inline modules/ftui/ftui.helper.js:232]: Please use === - Strict Equality Comparison

File: modules/ftui/ftui.helper.js [hunk 6]
```diff
@@ -227,6 +226,66 @@ export function durationFromSeconds(time) {
...
+
+  const SECONDS = ~~(x);
+  const seconds = ~~(x % 60);
+  x /= 60;
+  const MINUTES = ~~(x);
+  const minutes = ~~(x % 60);
+  x /= 60;
+  const HOURS = ~~(x);
```
Comment (knowthelist) [inline modules/ftui/ftui.helper.js:240]: Please only use upper case names for constants like PI = 3.1415; or MINUTES = 60, not for variables.

https://www.w3schools.com/js/js_conventions.asp

File: modules/ftui/ftui.binding.js (hunk 4 above)
Comment (OdfFhem) [reply modules/ftui/ftui.binding.js]: imode means inputMode.
fmode means formatMode.

Replaced

File: modules/ftui/ftui.helper.js (hunk 5 above)
Comment (OdfFhem) [reply modules/ftui/ftui.helper.js]: Two times replaced

File: modules/ftui/ftui.helper.js (hunk 6 above)
Comment (OdfFhem) [reply modules/ftui/ftui.helper.js]: Now using totalSeconds, totalMinutes, totalHours, totalDays.

**Pull request merged**